### PR TITLE
fix: sort bom components and dependencies

### PR DESCRIPTION
**Pull request by jeremylong**
While there is absolutely no requirement to do this - it makes testing easier (at least for me). As you make simple changes in the plugin or project using the plugin by sorting the components and dependencies it makes simple file compares much easier.

### CODE REVIEW

File: src/main/java/org/cyclonedx/gradle/utils/CycloneDxUtils.java [hunk 1]
```diff
@@ -82,6 +83,51 @@ public static void writeBom(
         }
     }
 
+    /**
+     * Utility method to sort the components and dependencies in the BOM.
+     * @param bom the BOM to sort
+     */
+    private static void sortBom(final Bom bom) {
```
[the remark this answers is not in the document]
Comment (jeremylong) [reply src/main/java/org/cyclonedx/gradle/utils/CycloneDxUtils.java:90]: The reason for the sorting is that there is no ordering. Changing a dependency in the project being scanned can vastly change the order in which dependencies are shown - as such I would love to add ordering. Other cycloneDx plugins [sort the dependencies](https://github.com/search?q=repo%3ACycloneDX%2Fcyclonedx-cocoapods%20sort&type=code) as well.

If others want a configuration option I can add it. However, I do not believe it is needed. I'll let @skhokhlov decide.

File: src/main/java/org/cyclonedx/gradle/utils/CycloneDxUtils.java
[the remark this answers is not in the document]
Comment (jeremylong) [reply src/main/java/org/cyclonedx/gradle/utils/CycloneDxUtils.java:1]: This is just sorting - I didn't think this needed an explicit test.

File: src/main/java/org/cyclonedx/gradle/utils/CycloneDxUtils.java [hunk 2]
```diff
@@ -71,6 +71,7 @@ public static void writeBom(
             final Version version,
             final String formats) {
 
+        sortBom(bom);
```
Comment (skhokhlov) [inline src/main/java/org/cyclonedx/gradle/utils/CycloneDxUtils.java:74]: Components and dependencies in the BOM are already sorted by using TreeSet here https://github.com/CycloneDX/cyclonedx-gradle-plugin/blob/ba5369b4f3792b9fcaf3237f26f75b0cd7014c53/src/main/java/org/cyclonedx/gradle/SbomBuilder.java#L95-L96. 

If you want to change the order, it must be done [here for components](https://github.com/CycloneDX/cyclonedx-gradle-plugin/blob/ba5369b4f3792b9fcaf3237f26f75b0cd7014c53/src/main/java/org/cyclonedx/gradle/model/ComponentComparator.java) and [here for dependencies](https://github.com/CycloneDX/cyclonedx-gradle-plugin/blob/ba5369b4f3792b9fcaf3237f26f75b0cd7014c53/src/main/java/org/cyclonedx/gradle/model/DependencyComparator.java)

File: src/main/java/org/cyclonedx/gradle/utils/CycloneDxUtils.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (skhokhlov) [reply src/main/java/org/cyclonedx/gradle/utils/CycloneDxUtils.java:90]: We already use TreeSet for components and dependencies, so there will be no performance degradation if you keep using it.

**Review by skhokhlov — CHANGES_REQUESTED** [review]
I believe that sorting components based on purl instead of bom-ref makes sense, but me done when plugin is collecting components, not afterwards.

**Comment by jeremylong** [comment]
I'm not sure how feasible sorting on insert would be updating just the plugin. The lists we are sorting are contained in `org.cyclonedx.model` which is provided by [cyclonedx-core-java](https://github.com/CycloneDX/cyclonedx-core-java). To sort on insert we would need to update the core-java implementation to use sorted lists and and implement comparable on the types to be sorted. In addition, just from a performance standpoint sorting afterwards is slightly faster.

@skhokhlov let me know if you want me to close this PR and open one against the cyclonedx-core-java library.

**Comment by skhokhlov** [comment]
@jeremylong it doesn't require changes for the core library because all the components are collected here with sorted sets and then they are just converted to regular lists, but the order is preserved. Please see my comment https://github.com/CycloneDX/cyclonedx-gradle-plugin/pull/549#discussion_r1881959696, it's just a single line change.

**Comment by jeremylong** [comment]
Sorry - this PR was opened based on the old 1.x implementation that did not sort the output. Closing this PR as I missed the use of TreeSort. I don't think we need to adjust from `bom-ref` to `purl` - as they are almost, if not always, equal.

**Pull request closed without merging**